### DISCUSSION TITLE
Groups claim not getting added

### DISCUSSION THREAD

**Discussion by dheeraj-326** [keycloak/keycloak#30113] [Q&A]
Version: 24.0.4

We are doing SSO login using Azure Entra ID and we want the groups claim to be mapped to the KC token from Azure's id token.

On Azure side we have it configured it like this, as their documentation instructs:
<img width="467" alt="image" src="https://github.com/keycloak/keycloak/assets/24998581/e965b21e-866d-4a14-a257-0b3dcfa25668">

On Keycloak side, we have:

1. IDP mapper defined as follows  

<img width="808" alt="image" src="https://github.com/keycloak/keycloak/assets/24998581/861423a0-77f4-4ffa-ac54-cb293655a9db">  

2. A client scope defined which has this mapper in it  

<img width="554" alt="image" src="https://github.com/keycloak/keycloak/assets/24998581/8b7617f9-6c2f-482a-94f8-9180de6b2d1f">

3. Client scope added to our public client  

<img width="645" alt="image" src="https://github.com/keycloak/keycloak/assets/24998581/7d051eb6-3d12-4739-845e-47e6fe5c5091">

But despite all this, KC token does not contain the groups claim. I believe I am missing some "magic configuration" which will finally make the claim appear. Any help us highly appreciated.

**Comment by danielFesenmeyer** [comment]
Your IDP mapper looks a bit suspicious. In Keycloak 23, there is an additional config value "User Attribute Name", which is either not shown in your screenshot or somehow does not exist in your setup. ("User Attribute Name" should be "groups" in your case.)

Btw, have you to tried to login at Entra ID directly and checked that groups are actually contained in the token? Furthermore, there is an issue in case of many groups in Entra ID: If a user has many groups, only a shortened claim is returned, with a link to all groups.

Maybe you also like to have a look at this discourse discussion, which provides some more insights on this topic: https://keycloak.discourse.group/t/groups-from-azure-ad/4876/7

**Reply by dheeraj-326** [reply]
Thanks @danielFesenmeyer for your response. I believe that the User Attribute Name was omitted in newer versions. Although we did suspect that it might be something related to it and we did try this with version 22.0.5 as well where also it didn't work (claim missing).

I did enable "Store Tokens" in our KC Idp and looked at the Azure tokens. We are seeing groups claim in the id_token but not the access_token. But we believe that this is not the cause because our stage instance of KC (version 22.0.5) is mapping these correctly (also the reason why I tried the above mappers with 22.0.5). It is just that the dev who configured it last year has forgotten that "magic configuration" which is what is different from stage and dev.
And regarding Azure skipping groups and adding a link to them, is not the case here which we have checked.

Again, thanks for trying to help. I will read through that article and try those steps. Will let you know if that worked.

**Reply by dheeraj-326** [reply]
I checked the answer you had linked to. On step 5, it talks about specifically about mapping a particular groupId to a Keycloak Role. This is not what we want. We want any groups present to be added as a claim in our token and that to be handled by our application logic as it sees fit. Do you think that can be done?

**Reply by danielFesenmeyer** [reply]
We have actually added all groups to the token with Keycloak 23, but with a custom identity provider implementation. Afaik, it was also possible with the default OIDC provider implementation, we just extended it to read all groups, in case there are too many for a user.
Sadly, I can't see what's different in your configuration.

What about your logs, do you get any warnings there?

**Reply by dheeraj-326** [reply]
Nothing unusual in the logs. We also would like to add groups automatically as there are too many. Is your custom implementation open source?

**Reply by danielFesenmeyer** [reply]
We have planned to contribute that to Keycloak, but haven't got the resources to do it yet.
You could join the discussion of this (rejected) PR (and the associated issue), were this topic is further discussed: https://github.com/keycloak/keycloak/pull/26193

**Reply by dheeraj-326** [reply]
It looks like you didn't add the links of the first PR.
